maro/cli/process/agent/job_agent.py

```diff
@@ -0,0 +1,169 @@
+            parallel_level = self.redis_connection.hget(ProcessRedisName.SETTING, "parallel_level")
+            # Start pending job only if current running job's number less than parallel level.
+            if int(parallel_level) > running_jobs_length:
+                self._start_job(job_detail)
+                self.redis_connection.lrem(ProcessRedisName.PENDING_JOB_TICKETS, 0, job_name)
+
+    def _start_job(self, job_details: dict):
+        command_pid_list = []
+        for component_type, command_info in job_details["components"].items():
+            component_number = command_info["num"]
+            component_command = command_info["command"]
+            for number in range(component_number):
+                job_local_path = os.path.expanduser(f"{LocalPaths.MARO_PROCESS}/{job_details['name']}")
+                if not os.path.exists(job_local_path):
+                    os.makedirs(job_local_path)
+
+                with open(f"{job_local_path}/{component_type}_{number}.log", "w") as log_file:
+                    proc = subprocess.Popen(component_command, shell=True, stdout=log_file)
+                    command_pid = get_child_pid(proc.pid)
+                    command_pid_list.append(command_pid)
+
+        self.redis_connection.hset(ProcessRedisName.RUNNING_JOB, job_details["name"], json.dumps(command_pid_list))
+
+
+class JobTrackingAgent(mp.Process):
+    def __init__(self, redis_connection, check_interval: int = 60):
+        super().__init__()
+        self.redis_connection = redis_connection
+        self.check_interval = check_interval
+        self._shutdown_count = 0
+
+    def run(self):
+        while True:
+            self._check_job_status()
+            time.sleep(self.check_interval)
+            keep_alive = int(self.redis_connection.hget(ProcessRedisName.SETTING, "keep_agent_alive"))
+            if not keep_alive:
+                self._close_agents()
+
+    def _check_job_status(self):
+        running_jobs = self.redis_connection.hgetall(ProcessRedisName.RUNNING_JOB)
+        running_jobs = {job_name.decode(): json.loads(pid_list) for job_name, pid_list in running_jobs.items()}
+
+        for running_job, pid_list in running_jobs.items():
+            # Check pid status
+            still_alive = False
+            for pid in pid_list:
+                if psutil.pid_exists(pid):
+                    still_alive = True
+
+            # Update if no pid exists
+            if not still_alive:
+                self.redis_connection.hdel(ProcessRedisName.RUNNING_JOB, running_job)
+
+    def _close_agents(self):
+        if (
+            not self.redis_connection.hlen(ProcessRedisName.RUNNING_JOB) and
+            not self.redis_connection.llen(ProcessRedisName.PENDING_JOB_TICKETS)
+        ):
+            self._shutdown_count += 1
+        else:
+            self._shutdown_count = 0
+
+        if self._shutdown_count >= 5:
```

Comment:
why 5 here

Reply:
Just personal design, i will add it into process setting.